### PR TITLE
Settings

### PR DESCRIPTION
**Pull request by msquare**
* Applied new settings layout
* Added settings section to change password according to new layout

### CODE REVIEW

File: resources/lang/en_US/default.po [hunk 1]
```diff
@@ -181,8 +181,17 @@ msgstr "Message"
...
 
+msgid "settings.profile"
+msgstr "Profile"
+
+msgid "settings.password"
+msgstr "Password"
+
 msgid "settings.oauth"
-msgstr "OAuth Settings"
+msgstr "Single-Sign-On"
```
Comment (MyIgel) [inline resources/lang/en_US/default.po:191]: Same here

File: resources/views/macros/form.twig [hunk 2]
```diff
@@ -77,3 +77,8 @@
 {% macro submit(label, opt) %}
     {{ _self.button(label|default(__('form.submit')), opt|default({})|merge({'type': 'submit'})) }}
 {%- endmacro %}
+
+{% macro info(text) %}
+{% import 'macros/base.twig' as m %}
```
Comment (MyIgel) [inline resources/views/macros/form.twig:82]: Hmm, imho it should be imported at the top? And i would also move it to the base macros as its not that form specific imho

File: resources/views/pages/news/edit.twig [hunk 3]
```diff
@@ -62,7 +62,7 @@
             {% if news %}
                 <div class="row">
                     <div class="col-md-12">
-                        <h2>Preview</h2>
+                        <h2>{{ __('Preview') }}</h2>
```
Comment (MyIgel) [inline resources/views/pages/news/edit.twig:65]: Oof, that should be a `news.preview` :D

File: includes/view/User_view.php [hunk 4]
```diff
@@ -37,23 +37,22 @@ function User_settings_view(
...
-    foreach ($oauth2_providers as $name => $config) {
-        $showOauth = $showOauth || !isset($config['hidden']) || !$config['hidden'];
-    }
-
     return page_with_title(settings_title(), [
         msg(),
         div('row', [
-            div('col-md-6', [
+            div('col-md-3 settings-menu', [
+                '<ul class="nav nav-pills nav-stacked">',
```
Comment (MyIgel) [inline includes/view/User_view.php:44]: That means that this navigation has to be changed in two places when editing it which is not that great, there has to be a better option using the twig template and adding the content directly to it

File: resources/views/pages/settings/password.twig [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+{% block title %}{{ __('settings.password') }}{% endblock %}
+
+{% block row_content %}
+        <form action="" enctype="multipart/form-data" method="post">
+            {{ csrf() }}
+
+            <div class="row">
+                <div class="col-md-12">
+                    {{ f.info(__('Here you can change your password.')) }}
```
Comment (MyIgel) [inline resources/views/pages/settings/password.twig:13]: These messages should be replaced by for example `settings.change-password`, `settings.old-password` etc

File: src/Controllers/SettingsController.php [hunk 6]
```diff
@@ -21,11 +34,54 @@ class SettingsController extends BaseController
...
+     */
+    public function savePassword(Request $request): Response
+    {
+        $user = $this->auth->user();
+
+        if (
+            !$request->has('password')
+            || !$this->auth->verifyPassword($user, $request->postData('password'))
+        ) {
+            $this->addNotification('-> not OK. Please try again.', 'errors');
```
Comment (MyIgel) [inline src/Controllers/SettingsController.php:72]: These should definitely be translation strings (and esp. this message should be reworded as its not that descriptive)

File: src/Controllers/SettingsController.php [hunk 7]
```diff
@@ -21,11 +34,54 @@ class SettingsController extends BaseController
...
     public function __construct(
+        Authenticator $auth,
         Config $config,
+        LoggerInterface $log,
+        Redirector $redirector,
         Response $response
     ) {
-        $this->config = $config;
-        $this->response = $response;
+            $this->auth = $auth;
```
Comment (MyIgel) [inline src/Controllers/SettingsController.php:43]: Interesting formatting which looks odd when comparing it tot he rest of the file

File: src/Controllers/SettingsController.php [hunk 8]
```diff
@@ -21,11 +34,54 @@ class SettingsController extends BaseController
...
+        return $this->response->withView(
+            'pages/settings/password.twig',
+            $this->getNotifications()
+        );
+    }
+
+    /**
+     * @return Response
+     */
+    public function savePassword(Request $request): Response
```
Comment (MyIgel) [inline src/Controllers/SettingsController.php:64]: I'm missing phpdoc tags for the Parameter ;)

File: src/Controllers/SettingsController.php [hunk 9]
```diff
@@ -21,11 +34,54 @@ class SettingsController extends BaseController
...
+    {
+        $user = $this->auth->user();
+
+        if (
+            !$request->has('password')
+            || !$this->auth->verifyPassword($user, $request->postData('password'))
+        ) {
+            $this->addNotification('-> not OK. Please try again.', 'errors');
+        } elseif (strlen($request->postData('new_password')) < config('min_password_length')) {
+            $this->addNotification('Your password is to short (please use at least 6 characters).', 'errors');
```
Comment (MyIgel) [inline src/Controllers/SettingsController.php:74]: The number should be replaced with a `%d` here to show the real configured value

File: tests/Unit/Controllers/SettingsControllerTest.php [hunk 10]
```diff
@@ -10,34 +10,237 @@
...
+    protected $user;
+
+    /**
+     * @covers \Engelsystem\Controllers\SettingsController::password
+     */
+    public function testPassword()
+    {
+        /** @var Response|MockObject $response */
+        $this->response->expects($this->once())
+        ->method('withView')
```
Comment (MyIgel) [inline tests/Unit/Controllers/SettingsControllerTest.php:52]: These should be indented by 4 spaces to be better readable, will have a look at how the linter can find those ;)

File: tests/Unit/Controllers/SettingsControllerTest.php [hunk 11]
```diff
@@ -10,34 +10,237 @@
...
+    public function testSavePassword()
+    {
+        $body = [
+            'password' => 'password',
+            'new_password' => 'newpassword',
+            'new_password2' => 'newpassword'
+        ];
+        $this->request = $this->request->withParsedBody($body);
+
+        $this->auth->expects($this->once())
```
Comment (MyIgel) [inline tests/Unit/Controllers/SettingsControllerTest.php:76]: :information_source: There is a `$this->setExpects(...)` method which shrinks it to one line

File: src/Controllers/SettingsController.php [hunk 12]
```diff
@@ -21,11 +34,54 @@ class SettingsController extends BaseController
...
+    }
+
+    /**
+     * @return Response
+     */
+    public function savePassword(Request $request): Response
+    {
+        $user = $this->auth->user();
+
+        if (
```
Comment (MyIgel) [inline src/Controllers/SettingsController.php:68]: I would recommend using the internal validator here

```php
$minLength = config('min_password_length');
$data = $this->validate($request, [
    'password' => 'required',
    'new_password' => 'required|min:' . $minLength,
    'new_password2' => 'required'
]);
```

Eventually the is equal to can be validated, that has to be checked. I would definitely rename the field to password_confirmation ^^

File: resources/views/pages/settings/password.twig [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+                        'password',
+                        __('Old password:'),
+                        'password',
+                        {'required': true}
+                    ) }}
+                    {{ f.input(
+                        'new_password',
+                        __('New password:'),
+                        'password',
+                        {'required': true}
```
Comment (MyIgel) [inline resources/views/pages/settings/password.twig:24]: I would definitely also add a html `minlength` attribute here (as `min` opt) which gets configured by the config which should mean that one sentence has to be translated less :upside_down_face:

**Review by MyIgel — CHANGES_REQUESTED** [review]

File: resources/views/macros/form.twig (hunk 2 above)
Comment (msquare) [reply resources/views/macros/form.twig:82]: That didn't work. Any other ideas?

Comment (MyIgel) [reply resources/views/macros/form.twig:82]: 0o

Thats interesting (and looks more like a bug than a feature..)

But would be irrelevant if moved to the `base.twig` file

Comment (msquare) [reply resources/views/macros/form.twig:82]: According to https://stackoverflow.com/questions/9548601/calling-a-macro-inside-another-macro-in-twig, the import was actually the correct way. It's even needed when moved to `base.twig`.

File: src/Controllers/SettingsController.php (hunk 9 above)
Comment (msquare) [reply src/Controllers/SettingsController.php:74]: This is now done by Validator and browser input length.

File: resources/views/macros/form.twig (hunk 2 above)
Comment (MyIgel) [reply resources/views/macros/form.twig:82]: When referencing macros from the same file the `_self` keyword can be used ;)

File: resources/views/pages/settings/password.twig (hunk 13 above)
Comment (msquare) [reply resources/views/pages/settings/password.twig:24]: Added minlength to registration and password reset as well.

File: includes/view/User_view.php (hunk 4 above)
Comment (msquare) [reply includes/view/User_view.php:44]: Sorry, won't happen until settings are completely rewritten. Tried 5 things that didn't work.

Comment (MyIgel) [reply includes/view/User_view.php]: It works for me, see 35aff139f7c0d140cf1751b104fa2e5320ea6dba ;)

File: resources/lang/de_DE/additional.po [hunk 14]
```diff
@@ -27,7 +27,10 @@ msgid "validation.email.email"
 msgstr "Die E-Mail-Adresse ist nicht gültig."
 
 msgid "validation.password.min"
-msgstr "Dein angegebenes Passwort ist zu kurz."
+msgstr "Dein neues Passwort ist zu kurz."
```
Comment (MyIgel) [inline resources/lang/de_DE/additional.po:30]: That also affects the registration pssword but as it should be never used (the form has the minlnength requirement) the wording is not that important here :shrug:

File: resources/lang/de_DE/default.po [hunk 15]
```diff
@@ -238,11 +238,6 @@ msgid "We sent you an email containing your password recovery link."
 msgstr ""
 "Wir haben dir eine eMail mit einem Link zum Passwort-zurücksetzen geschickt."
 
-#: resources/views/pages/password/reset-success.twig:9
-#: includes/pages/user_settings.php:120
-msgid "Password saved."
```
Comment (MyIgel) [inline resources/lang/de_DE/default.po:243]: This string is used in `resources/views/pages/password/reset-success.twig`

File: src/Controllers/SettingsController.php [hunk 16]
```diff
@@ -39,10 +107,27 @@ public function oauth(): Response
...
         );
     }
+
+    /**
+     * @return array
+     */
+    public function settingsMenu(): array
+    {
+        $menu = [
+            url('/user-settings') => 'settings.profile',
```
Comment (MyIgel) [inline src/Controllers/SettingsController.php:124]: This should be moved to `additional.po` as it has no direct translation reference

**Review by MyIgel — CHANGES_REQUESTED** [review]

**Review by MyIgel — APPROVED** [review]

**Pull request merged**